### PR TITLE
Move magic "compiled" string behind an API in symbolt

### PR DESCRIPTION
**Pull request by tautschnig**

### CODE REVIEW

**Comment by tautschnig** [comment]
Factored out from #1987.

File: src/util/symbol.h [hunk 1]
```diff
@@ -107,6 +107,16 @@ class symbolt
   {
     return !is_type && !is_macro && type.id()==ID_code;
   }
+
+  bool is_compiled() const
```
Comment (smowton) [inline src/util/symbol.h]: Document what this means here

Comment (tautschnig) [reply src/util/symbol.h]: Documentation (and additional invariants) added.

**Review by smowton — APPROVED** [review]

**Review by kroening — APPROVED** [review]

File: src/util/symbol.h [hunk 2]
```diff
@@ -107,6 +108,21 @@ class symbolt
...
+  /// Returns true iff the the symbol's value has been compiled to a goto
+  /// program.
+  bool is_compiled() const
+  {
+    return type.id() == ID_code && value == exprt(ID_compiled);
+  }
+
+  /// Set the symbol's value to "compiled"; to be used once the code-typed value
+  /// has been converted to a goto program.
+  void mark_compiled()
```
Comment (peterschrammel) [inline src/util/symbol.h]: Maybe better `set_compiled`?

**Comment by tautschnig** [comment]
Travis is currently failing with what looks like an instance of https://github.com/travis-ci/travis-ci/issues/9361. I'll re-try in the morning.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 9082573ffe5e2a3bbffcf78c70aed91b4fb99150).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85254753
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 637b17bc4e0c0ccde8d6138cb8a4b0f494830482).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85259736

**Pull request merged**